fdbserver/storageserver.actor.cpp

```
@@ -2470,6 +2509,440 @@ ACTOR Future<Void> getKeyValuesQ(StorageServer* data, GetKeyValuesRequest req)
+	// Lazily parse key and/or value to tuple because they may not need to be a tuple if not used.
+	Optional<Tuple> keyTuple;
+	Optional<Tuple> valueTuple;
+
+	Tuple hopKeyTuple;
+	for (int i = 0; i < hopKeyFormatTuple.size(); i++) {
+		Tuple::ElementType type = hopKeyFormatTuple.getType(i);
+		if (type == Tuple::BYTES || type == Tuple::UTF8) {
+			std::string s = hopKeyFormatTuple.getString(i).toString();
+			auto sz = s.size();
+
+			// Handle escape.
+			bool escaped = false;
+			size_t p = 0;
+			while (true) {
+				size_t found = s.find("{{", p);
+				if (found == std::string::npos) {
+					break;
+				}
+				s.replace(found, 2, "{");
+				p += 1;
+				escaped = true;
+			}
+			p = 0;
+			while (true) {
+				size_t found = s.find("}}", p);
+				if (found == std::string::npos) {
+					break;
+				}
+				s.replace(found, 2, "}");
+				p += 1;
+				escaped = true;
+			}
+			if (escaped) {
+				// If the element uses escape, cope the escaped version.
+				hopKeyTuple.append(s);
+			}
+			// {K[??]} or {V[??]}
+			else if (sz > 5 && s[0] == '{' && (s[1] == 'K' || s[1] == 'V') && s[2] == '[' && s[sz - 2] == ']' &&
+			         s[sz - 1] == '}') {
+				int idx;
+				try {
+					idx = std::stoi(s.substr(3, sz - 5));
+				} catch (std::exception& e) {
+					throw hop_bad_index();
+				}
+				Tuple* referenceTuple;
+				if (s[1] == 'K') {
+					// Use keyTuple as reference.
+					if (!keyTuple.present()) {
+						// May throw exception if the key is not parsable as a tuple.
+						keyTuple = Tuple::unpack(keyValue->key);
+					}
+					referenceTuple = &keyTuple.get();
+				} else if (s[1] == 'V') {
+					// Use valueTuple as reference.
+					if (!valueTuple.present()) {
+						// May throw exception if the value is not parsable as a tuple.
+						valueTuple = Tuple::unpack(keyValue->value);
+					}
+					referenceTuple = &valueTuple.get();
+				} else {
+					ASSERT(false);
+					throw internal_error();
```

Can we throw an error with more info here? So that we can quickly identify where is the `internal_error()` thrown.

---

This piece of code will be removed in next PR, so I wish to no change it for now. https://github.com/nblintao/foundationdb/pull/2/commits/2b627b9815aa7c9e19fd1f3bc5eeb54b228cf8a5